adapters/grpc/document_grpc.go

```diff
@@ -0,0 +1,166 @@
+	}
+}
+
+func (s *DocumentServer) Get(ctx context.Context, req *pb.DocGetRequest) (*pb.DocGetResponse, error) {
+	if ok, err := s.checkPluginRegistered(); ok {
+		key := toSdkKey(req.Key)
+		subKey := toSdkKey(req.SubKey)
+
+		if val, err := s.documentPlugin.Get(*key, subKey); err == nil {
+			if valStruct, err := structpb.NewStruct(val); err == nil {
+				return &pb.DocGetResponse{
+					Value: valStruct,
+				}, nil
+			} else {
+				// Case: Failed to create PB struct from stored value
+				// TODO: Translate from internal Document Plugin Error
+				return nil, err
+			}
+		} else {
+			// Case: There was an error retrieving the keyvalue
+			// TODO: Translate from internal Document Plugin Error
+			return nil, err
+		}
+	} else {
+		// Case: The Document plugin was not registered
+		// TODO: Translate from internal Document Plugin Error
+		return nil, err
+	}
+}
+
+func (s *DocumentServer) Delete(ctx context.Context, req *pb.DocDeleteRequest) (*pb.DocDeleteResponse, error) {
+	if ok, err := s.checkPluginRegistered(); ok {
+		key := toSdkKey(req.Key)
+		subKey := toSdkKey(req.SubKey)
+
+		if err := s.documentPlugin.Delete(*key, subKey); err == nil {
+			return &pb.DocDeleteResponse{}, nil
+		} else {
+			// Case: Failed to create the keyvalue
+			// TODO: Translate from internal Document Plugin Error
+			return nil, err
+		}
+	} else {
+		// Case: Plugin was not registered
+		return nil, err
+	}
+}
+
+func (s *DocumentServer) Query(ctx context.Context, req *pb.DocQueryRequest) (*pb.DocQueryResponse, error) {
+	if ok, err := s.checkPluginRegistered(); ok {
+		key := toSdkKey(req.Key)
+		subcoll := req.GetSubCollection()
+		expressions := make([]sdk.QueryExpression, len(req.GetExpressions()))
+		for i, exp := range req.GetExpressions() {
+			expressions[i] = sdk.QueryExpression{
+				Operand:  exp.GetOperand(),
+				Operator: exp.GetOperator(),
+				Value:    exp.GetValue(),
+			}
+		}
+		limit := int(req.GetLimit())
+		pagingMap := req.GetPagingToken()
+
+		if qr, err := s.documentPlugin.Query(*key, subcoll, expressions, limit, pagingMap); err == nil {
+
+			valStructs := make([]*structpb.Struct, len(qr.Data))
+			for i, valMap := range qr.Data {
+				if valStruct, err := structpb.NewStruct(valMap); err == nil {
+					valStructs[i] = valStruct
+
+				} else {
+					// Case: Failed to create PB struct from stored value
+					// TODO: Translate from internal Document Plugin Error
+					return nil, err
+				}
+			}
+
+			return &pb.DocQueryResponse{
+				Values:      valStructs,
+				PagingToken: qr.PagingToken,
+			}, nil
+
+		} else {
+			// Case: Failed to create the keyvalue
+			// TODO: Translate from internal Document Plugin Error
+			return nil, err
+		}
+
+	} else {
+		// Case: Plugin was not registered
+		return nil, err
+	}
+}
+
+func NewDocumentServer(docPlugin sdk.DocumentService) pb.DocumentServer {
+	return &DocumentServer{
+		documentPlugin: docPlugin,
+	}
+}
+
+func toSdkKey(key *pb.Key) *sdk.Key {
+	if key != nil {
```

Comment:
Can add Key validation at the adapter layer here (or in a separate method) to universally validate non-nil keys.